citizennet/purescript-ocelot
# Overhaul how typeaheads function

## What does this pull request do?

This pull request is a major overhaul of how typeaheads behave and expose queries. It relies on PR 18 in the `select` library:

https://github.com/citizennet/purescript-halogen-select/pull/18/

**Motivations**

- The `load` function is tedious and requires too much involvement from the parent. If the parent has to load all the data constantly, why have encapsulated typeaheads? This PR allows typeaheads to fetch their own data internally without involvement from the parent. Since they already have to carry an effects row around to support debouncing, the only added cost is the requirement to support `AJAX` also.

- It is too difficult to update items in typeaheads -- to the point where sometimes we have to send an entire input again. This update supports `ReplaceSelections` and `ReplaceItems` for easy access to the typeahead items and selections.

- It's annoying to have to query a typeahead for its selections if you want to watch them on the parent. This supports the ability to raise messages when an item is selected or removed in a typeahead so the parent can track this and save it on its own state.

- ContinuousAsync typeaheads now run without requiring any input from the parent, which can instead just watch for `SelectionsChanged` messages from the component.

- There doesn't seem to be a case where we want a single-run async typeahead, so this update removes them. **However, we may want to replace it as a triggered-fetch async -- AKA "When I tell you, fetch your data using your provided function."** This may be worth it now that we can embed our loading functions within the typeahead.

- This incorporates several fixes to how search behaviors work, bubbling up the work done in `select`.


## Where should the reviewer start?

The most important files to check out are:

1. The new `Core` typeahead module, which is much simpler and cleaner: https://github.com/citizennet/purescript-ocelot/compare/search-or-toggle?expand=1#diff-c3eb2fb0cca5780c1528f35104090940

2. The new `Component` typeahead module. Sync typeaheads are almost entirely unchanged, whereas async typeaheads now take an input function to run. https://github.com/citizennet/purescript-ocelot/compare/search-or-toggle?expand=1#diff-406150f1b430c9d6bfe7f71613e0e4b3

3. The `Validation` module, where you can see that the typeahead types are much less scary, and that the added burden in the input record is minimal, whereas the savings in the `load` handling more than make up for it. This also illustrates how we might incrementally trigger a typeahead from an action taken on another one. https://github.com/citizennet/purescript-ocelot/compare/search-or-toggle?expand=1#diff-526ae140cab1fde6792879a2a4067d50

## How should this be manually tested?

Build the project, then open the `Validation` module to verify behaviors work as intended. Try searching Developers, then select something from the last typeahead, then try searching again to see the incremental result.

## Other Notes:

Wildcat will require an update to use this work. Performing that update consists only of updating the 5 typeahead inputs, which in turn is surprisingly minimal as demonstrated by the `Validation` changes below:

https://github.com/citizennet/purescript-ocelot/compare/search-or-toggle?expand=1#diff-526ae140cab1fde6792879a2a4067d50

- The input types won't need to change
- We can replace the massive update records with `ReplaceItems` and `ReplaceSelections`
- We can watch for `SelectionsChanged` and attach a state modifier to the event

I'd estimate ~1hr of work to make the update.
